internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -757,3 +784,37 @@ func EnqueueForCompositionRevisionFunc(of resource.CompositeKind, list func(ctx
 		}
 	}
 }
+
+// handleFunctionResults will process the result returned by the composition functions.
+// This includes recording events for the xr and claim and setting updating status conditions
+// on the xr.
+func handleFunctionResults(r *Reconciler, cnd *conditions.ConditionProcessor, log logging.Logger, cm *claim.Unstructured, xr *composite.Unstructured, res CompositionResult) {
+	// record all events
+	for t, es := range res.Events {
+		for _, e := range es {
+			log.Debug(e.Message)
+			if xr != nil {
+				r.record.Event(xr, e)
+			}
+			if t == v1beta1.Target_TARGET_COMPOSITE_AND_CLAIM && cm != nil {
+				r.record.Event(cm, e)
+			}
```

Comment:
I'd prefer to avoid the XR controller emitting claim events if possible.

Currently the XR controller is mostly (entirely?) ignorant of claims. The claim controller is entirely responsible for syncing claims with XRs. I'd like to maintain that separation of concerns.

Events are really [just Kubernetes objects](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/). So one way to absolve the XR controller from emitting claim events would be to:

* Have the XR controller label events with their target - e.g. `target.crossplane.io: composite-and-claim`
* Add a new controller that watches for events with the `target.crossplane.io: composite-and-claim` label and replicates those events to the claim

I think the only unfortunate part about this approach would be that we'd need to trim the prefix that contains information about what pipeline step produced the result. I think I still prefer it for the separation of concern, though.

Reply:
Another way to allow the claim reconciler to emit the events would be to update the `claimConditions` to contain the event type.
```yaml
claimConditions:
- conditionType: DatabaseReady
  eventType: Warning
```

IDK if this would be preferred, just came to mind.